lib/Transforms/LayoutOptimization/LayoutOptimization.cpp

```diff
@@ -0,0 +1,319 @@
+  if (resultLayoutConversions.empty()) {
+    LLVM_DEBUG(
+        llvm::dbgs()
+        << "Skipping op, no results needed immediate layout conversion\n");
+    return success();
+  }
+
+  // Now compute the cost of hoisting each conversion layout.
+  SmallVector<HoistingResult> hoistingResults;
+  for (auto resultLayoutConversion : resultLayoutConversions) {
+    auto result = computeHoistedCost(op, resultLayoutConversion);
+    LLVM_DEBUG(llvm::dbgs() << "\tHoisting layout " << result.newOutputLayout
+                            << " will cost " << result.cost << "\n");
+    hoistingResults.push_back(result);
+  }
+
+  // For each result layout conversion, try to hoist above the op, get the
+  // minimum cost
+  auto minHoistingResult = llvm::min_element(
+      hoistingResults, [](const HoistingResult &a, const HoistingResult &b) {
+        return a.cost < b.cost;
+      });
+
+  if (minHoistingResult->cost > 0) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Skipping op, no hoisting result with positive cost\n");
+    return success();
+  }
+
+  LLVM_DEBUG(llvm::dbgs()
+             << "Minimum cost layout conversion converts output layout from "
+             << op->getAttr(tensor_ext::TensorExtDialect::kLayoutAttrName)
+             << " to layout " << minHoistingResult->newOutputLayout << "\n");
+
+  builder.setInsertionPoint(op);
+  for (auto i = 0; i < op->getNumOperands(); ++i) {
+    // Convert each operand layout given the result.
+    LLVM_DEBUG(llvm::dbgs()
+               << "Converting operand " << i << " with layout "
+               << findAttributeAssociatedWith(
+                      op->getOperand(i),
+                      tensor_ext::TensorExtDialect::kLayoutAttrName)
+               << " to layout " << minHoistingResult->newInputLayouts[i]
+               << "\n");
+    auto operand = op->getOperand(i);
+    auto originalLayout = findAttributeAssociatedWith(
+        operand, tensor_ext::TensorExtDialect::kLayoutAttrName);
+    if (failed(originalLayout)) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Failed to get layout for operand : " << i << "\n");
+      return failure();
+    }
+    AffineMapAttr newInputLayout =
+        AffineMapAttr::get(minHoistingResult->newInputLayouts[i]);
+    auto newInput = builder.create<ConvertLayoutOp>(
+        op->getLoc(), operand, cast<AffineMapAttr>(originalLayout.value()),
+        newInputLayout);
+    newInput->setAttr(tensor_ext::TensorExtDialect::kLayoutAttrName,
+                      newInputLayout);
+    builder.replaceUsesWithIf(operand, newInput, [&](OpOperand &operand) {
+      return operand.getOwner() == op;
+    });
+  }
+
+  // Set new layout attribute for the op.
+  SmallVector<AffineMap> newOutputLayouts(op->getNumResults(),
+                                          minHoistingResult->newOutputLayout);
+  op->setAttr(tensor_ext::TensorExtDialect::kLayoutAttrName,
+              builder.getAffineMapArrayAttr(newOutputLayouts));
```

Comment:
```suggestion
  op->setAttr(tensor_ext::TensorExtDialect::kLayoutAttrName,
              minHoistingResult->newOutputLayout);
```

The op only has one result, as checked above.